skyportal/utils/gcn.py

```diff
@@ -0,0 +1,254 @@
+from astropy.coordinates import ICRS
+from astropy_healpix import HEALPix, nside_to_level, pixel_resolution_to_nside
+import ligo.skymap.io
+import ligo.skymap.postprocess
+import ligo.skymap.moc
+
+
+def get_dateobs(root):
+    """Get the UTC event time from a GCN notice, rounded to the nearest second,
+    as a datetime.datetime object."""
+    dateobs = Time(
+        root.find(
+            "./WhereWhen/{*}ObsDataLocation"
+            "/{*}ObservationLocation"
+            "/{*}AstroCoords"
+            "[@coord_system_id='UTC-FK5-GEO']"
+            "/Time/TimeInstant/ISOTime"
+        ).text,
+        precision=0,
+    )
+
+    # FIXME: https://github.com/astropy/astropy/issues/7179
+    dateobs = Time(dateobs.iso)
+
+    return dateobs.datetime
+
+
+def get_tags(root):
+    """Get source classification tag strings from GCN notice."""
+    # Get event stream.
+    mission = urlparse(root.attrib['ivorn']).path.lstrip('/')
+    yield mission
+
+    # What type of burst is this: GRB or GW?
+    try:
+        value = root.find("./Why/Inference/Concept").text
+    except AttributeError:
+        pass
+    else:
+        if value == 'process.variation.burst;em.gamma':
+            yield 'GRB'
+        elif value == 'process.variation.trans;em.gamma':
+            yield 'transient'
+
+    # LIGO/Virgo alerts don't provide the Why/Inference/Concept tag,
+    # so let's just identify it as a GW event based on the notice type.
+    notice_type = gcn.get_notice_type(root)
+    if notice_type in {
+        gcn.NoticeType.LVC_PRELIMINARY,
+        gcn.NoticeType.LVC_INITIAL,
+        gcn.NoticeType.LVC_UPDATE,
+        gcn.NoticeType.LVC_RETRACTION,
+    }:
+        yield 'GW'
+
+    # Is this a retracted LIGO/Virgo event?
+    if notice_type == gcn.NoticeType.LVC_RETRACTION:
+        yield 'retracted'
+
+    # Is this a short GRB, or a long GRB?
+    try:
+        value = root.find(".//Param[@name='Long_short']").attrib['value']
+    except AttributeError:
+        pass
+    else:
+        if value != 'unknown':
+            yield value.lower()
+
+    # Gaaaaaah! Alerts of type FERMI_GBM_SUBTHRESH store the
+    # classification in a different property!
+    try:
+        value = root.find(".//Param[@name='Duration_class']").attrib['value'].title()
+    except AttributeError:
+        pass
+    else:
+        if value != 'unknown':
+            yield value.lower()
+
+    # Get LIGO/Virgo source classification, if present.
+    classifications = [
+        (float(elem.attrib['value']), elem.attrib['name'])
+        for elem in root.iterfind("./What/Group[@type='Classification']/Param")
+    ]
+    if classifications:
+        _, classification = max(classifications)
+        yield classification
+
+    search = root.find("./What/Param[@name='Search']")
+    if search is not None:
+        yield search.attrib['value']
+
+
+def get_skymap(root, gcn_notice):
+    mission = urlparse(root.attrib['ivorn']).path.lstrip('/')
+
+    # Try Fermi GBM convention
```

Comment:
Is there a particular order in which these different conventions should be tested? Obviously I know nothing about the conventions or the format of the skymaps, but just wanted to make sure there weren't potentially cases where you would want to test for LVC before Fermi or whatever, since it's rare programmatically to have a string of `if` cases like this without any `else if`s

Reply:
No they are all separate (LVC, Fermi, neutrino, etc.). So I think its all the same.